[Feature] Table Child Rows

grahamhency opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:
<!-- Please check all that apply using "x". -->
- [x] The commit message follows our [guidelines](https://github.com/mobi/goponents/blob/master/CONTRIBUTING.md)
- [ ] Tests for the changes have been added
- [x] Docs have been added or updated

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [x] Style Update (CSS)
- [ ] Other... Please describe:

## What is the current behavior?
None

## What is the new behavior?
Adds ability to define child rows from a parent row. This differs from a detail row in that it embeds multiple rows depending on the number of children instead of a single row.

Resolves #672, based off of #679 

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.html:
```diff
@@ -123,18 +110,43 @@ <h3 class="go-heading-3 go-heading--no-margin go-table-header__title" *ngIf="tab
+                class="go-table__body-column"
+                width="60">
+              </td>
+              <td
+                *ngFor="let col of childRowColumns"
+                class="go-table__body-column"
+                [ngStyle]="{ 'vertical-align': col.alignment }">
```
Here we loop over each column declared by the implementor, exactly as we do for the regular `go-table-column`s.

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.html:
```diff
@@ -123,18 +110,43 @@ <h3 class="go-heading-3 go-heading--no-margin go-table-header__title" *ngIf="tab
+                *ngFor="let col of childRowColumns"
+                class="go-table__body-column"
+                [ngStyle]="{ 'vertical-align': col.alignment }">
+                <ng-container *ngIf="!col.goTableCell; else childColumnCellOutlet">
+                  {{ col.getFieldData(childItem) }}
+                </ng-container>
```
If the specified column does not have a custom template declared, then we output the value for the column.

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.html:
```diff
@@ -123,18 +110,43 @@ <h3 class="go-heading-3 go-heading--no-margin go-table-header__title" *ngIf="tab
+                <ng-container *ngIf="!col.goTableCell; else childColumnCellOutlet">
+                  {{ col.getFieldData(childItem) }}
+                </ng-container>
+                <ng-template #childColumnCellOutlet>
+                  <ng-container *ngTemplateOutlet="col.goTableCell context: { $implicit: childItem, parentItem: item }">
+                  </ng-container>
+                </ng-template>
```
If the specified column _does_ have a custom template defined, then we render that here and pass the entire `childItem` object implicitly and the entire `parentItem` object as a static variable.

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.scss:
```diff
@@ -288,32 +293,27 @@
 }
+
 // Hierarchical Table
-.hierarchical-go-table__body-row:first-child {
-  border-top: 1px solid $theme-light-app-bg;
-}
-.hierarchical-go-table__body-row:last-child {
-  border-bottom: 0;
-}
-.fixed-even-row {
-  background: #ffff;
-}
-.fixed-odd-row {
-  background: rgba($theme-light-app-bg, 0.5);
-}
-.active-parent-row {
-  background-color: #f6f6f6;
-  border-top: 1px solid #f0f0f0;
-}
-.active-row {
-  border-left: 4px solid #65b360;
-  padding: 5.7rem 0rem 0rem 0rem;
```
By going the route I am proposing, we wouldn't need most of these. Also, for your information, we use the [BEM CSS naming convention](http://getbem.com/introduction/) so I renamed these classes to follow that convention.

A couple other additional notes for writing CSS within Goponents:
- We always avoid using hard-coded values for colors, you can find all our color variables in `_variables.scss`
- We don't actually use the Tangoe brand color anywhere. When something needs to be branded as the "brand color" then we use the `GoConfigService` and get the `brandColor` for the current configuration

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.ts:
```diff
@@ -30,22 +39,24 @@ import {
+   * Used to dynamically access the child rows on each row
+   * item when implementing the goTableChildRows template
+   */
+  @Input() childRowsKey: string = null;
```
This is the key to accessing the children of the row I referred to.

We cannot assume that the child property of the data will be `children` for every set of data. This allows the implementor to specify the key of the property keeping the logic dynamic for every implementor.

grahamhency commented on projects/go-lib/src/lib/components/go-table/go-table.component.ts:
```diff
@@ -127,6 +149,12 @@ export class GoTableComponent implements OnInit, OnChanges, AfterViewInit {
     this.showTable = Boolean(this.tableConfig);
   }
 
+  setupConfigService(): void {
+    this.subscriptions.add(this.goConfigService.config.subscribe((config: GoConfigInterface) => {
+      this.brandColor = config.brandColor;
+    }));
+  }
```
This is where we `subscribe` to the `GoConfigService` and assign a local `brandColor` variable to that of the current configs.

jaredami replied:
I probably should have pointed this out to you earlier, but in [this PR](https://github.com/mobi/goponents/pull/693) I introduced a new method for unsubscribing when components are destroyed that utilizes the `takeUntil` RxJs operator.

So in this case, since this component has already been set up with that stuff, all you would have to do is change this to:
```javascript
this.goConfigService.config
  .pipe(takeUntil(this.destroy$))
  .subscribe((config: GoConfigInterface) => {
      this.brandColor = config.brandColor;
  });
```

And so you could remove `subscriptions: Subscription = new Subscription();` above.

grahamhency commented on projects/go-tester/src/app/components/off-canvas-test/off-canvas-test.component.html:
```diff
@@ -1,7 +1,4 @@
 <div class="go-container">
-  <go-off-canvas-header-actions>
-    <go-button>Close</go-button>
-  </go-off-canvas-header-actions>
```
This is breaking the build, it is fixed with a rebase to `dev`.. but I didn't want to do that in this PR.

grahamhency commented on projects/go-tester/src/app/components/test-page-1/test-page-1.component.html:
```diff
@@ -83,19 +84,24 @@
-      </tr>
-    </ng-template>
+
+    <ng-container #goTableChildRows>
+      <go-table-child-column field="id"></go-table-child-column>
+      <go-table-child-column field="name.first"></go-table-child-column>
+      <go-table-child-column field="name.last"></go-table-child-column>
+      <go-table-child-column field="email"></go-table-child-column>
+      <go-table-child-column field="gender"></go-table-child-column>
+      <go-table-child-column>
+        <ng-template #goTableCell let-childItem let-parentItem="parentItem">
+          {{ childItem.ip_address }}
+          <go-icon-button
+            buttonIcon="explore"
+            (handleClick)="logRow(parentItem, childItem)">
+          </go-icon-button>
+        </ng-template>
+      </go-table-child-column>
+    </ng-container>
```
This is what I am proposing for implementation. We specify a container for the child rows, then each row is represented by a `go-table-child-column` component in the exact way that the parent rows are defined.

In the last column, I pass a custom template and show how you could access both the current item and the parent item.

grahamhency commented on projects/go-tester/src/app/components/test-page-1/test-page-1.component.html:
```diff
@@ -70,6 +70,7 @@
   <!-- Hierarchical Table -->
   <go-table [loadingData]="tableDetailsLoading"
             [tableConfig]="tableDetailsConfig"
+            childRowsKey="children"
```
Here I am defining the key for the table to reference when looping through.

grahamhency commented on projects/go-tester/src/app/components/test-page-1/test-page-1.component.ts:
```diff
@@ -40,18 +39,16 @@ export class TestPage1Component implements OnInit {
-      this.tableLoading = false;
-    });
 
-    this.appService.getMockData(new GoTableConfig()).subscribe((data: any) => {
```
I removed a lot of this stuff just to clean it up a bit, we didn't really need to make this call twice.

grahamhency commented on projects/go-tester/src/app/components/test-page-1/test-page-1.component.ts:
```diff
@@ -95,4 +92,14 @@ export class TestPage1Component implements OnInit {
     this.toasterService.toastError({ message: 'Cancel clicked!' });
   }
+
+  logRow(parentItem: any, childItem: any): void {
+    this.toasterService.toastInfo({
+      header: 'Child Row Clicked!',
+      message: `
+        Child Name: ${childItem.name.first}<br/>
+        Parent Name: ${parentItem.name.first}
+      `
+    });
+  }
```
This is the method called from within the child row, you can see that you have access to both the specific child and parent objects.

jaredami commented on projects/go-style-guide/src/app/features/ui-kit/components/table-docs/components/table-child-rows/table-child-rows.component.html:
```diff
@@ -0,0 +1,95 @@
+        Sometimes within a table row we might need to show child rows related to the row. This is where the table child rows comes
+        into play. This provides a dropdown for each table row that when opened will render child rows as defined by the implementor.
+        With the table child rows template we allow the implementor to provide a template for each column to use for the row and then we inject
+        that template into the correct place in the interface, this works exactly like the normal table columns.
```
Just a minor grammatical note: should "this" be the start of a new sentence? Or perhaps it could be put in parentheses:

"With the table child rows template we allow the implementor to provide a template for each column to use for the row and then we inject that template into the correct place in the interface (this works exactly like the normal table columns)."

grahamhency replied:
Fair enough, I cleaned up the grammar here a bit.

jaredami reviewed (changes requested):
Just had a couple comments, but this is looking great. Implementation is nice and simple 👍

grahamhency commented:
@jaredami I didn't know about the `destroy` stuff so that's cool, thanks for informing me. I went ahead and removed the `subsciptions` variable and implemented the `takeUntil` approach. I also cleaned up the grammar on the docs.

Ready for a second review.

jaredami reviewed (approved):
Looks good to me 🔥
